drivers/counter/counter_ace_v1x_art.c

```diff
@@ -0,0 +1,129 @@
+/*
+ * Copyright (c) 2022 Intel Corporation
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <device.h>
+#include <devicetree.h>
+#include <zephyr/drivers/counter.h>
+#include <soc.h>
+#include <ace_v1x-regs.h>
+#include <counter/counter_ace_v1x_art_regs.h>
+
+#if defined(__XCC__)
+#include <xtensa/config/core.h>
+#if XCHAL_HAVE_HIFI4 && (defined(__ALDERLAKE__) \
+	|| defined(__METEORLAKE__) || defined(__METEORLAKE_S__))
+#include <xtensa/tie/xt_hifi4.h>
+#elif XCHAL_HAVE_HIFI3
+#include <xtensa/tie/xt_hifi3.h>
+#endif
+#endif
+
+static struct k_spinlock lock;
+
+#define ACE_TSCTRL	(0x72040)
+
+static void counter_ace_v1x_art_ionte_set(bool new_timestamp_enable)
+{
+	struct DfTSCTRL val;
+
+	val.u.full                      = ((volatile struct DfTSCTRL*)(ACE_TSCTRL))->u.full;
+	val.u.part.ionte                = new_timestamp_enable;
+	((volatile struct DfTSCTRL*)(ACE_TSCTRL))->u.full = val.u.full;
+}
+
+static void counter_ace_v1x_art_cdmas_set(uint32_t cdmas)
+{
+	struct DfTSCTRL val;
+
+	val.u.full                      = ((volatile struct DfTSCTRL*)(ACE_TSCTRL))->u.full;
```

Comment:
Please use sys_write32 and sys_read32 to access hardware registers.

Reply:
Actually other areas of intel_dsp code have an extensive history of using struct access to MMIO registers.  I actually really like that style, personally, but the traditional Zephyr API is obviously fine too.

But if you do this, the way it works is to put the volatile declaration in a e.g. your "ACE_TSCTRL" macro., so you just write `ACE_TSCTRL->u.full`.